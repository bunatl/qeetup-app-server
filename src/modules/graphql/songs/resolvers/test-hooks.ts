import { gql } from 'apollo-server';

export const SONG_FRAGMENT = gql`
    fragment SongFragment on Song {
        name
        artist
        cover
        description
        listens
        tags {
            value
            isImportant
        }
        audio
        isLiked
        comments {
            user {
                name
                avatar
                isArtist
            }
            text
        }
    }
`;
