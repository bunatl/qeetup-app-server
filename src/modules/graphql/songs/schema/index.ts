import { gql } from 'apollo-server';

export const typeDefs = gql`
    extend type Query {
        songs: [Song!]!
    }

    type Song {
        name: String!
        artist: String!
        cover: String!
        description: String!
        listens: Int!
        tags: [Tag!]!
        audio: String
        isLiked: Boolean
        comments: [Comment!]
    }

    type Tag {
        value: String!
        isImportant: Boolean
    }

    type Comment {
        user: User!
        text: String!
    }

    type User {
        name: String!
        avatar: String!
        isArtist: Boolean
    }
`;
