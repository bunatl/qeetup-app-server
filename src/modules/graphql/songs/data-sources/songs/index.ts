import { InMemoryDataSource } from '@container/data-sources/in-memory';
import { Song } from '@container/schema';

const songs: Song[] = [
    {
        artist: 'September 87',
        name: 'Bad Dream Baby',
        listens: 145202,
        isLiked: true,
        cover: 'https://i.vimeocdn.com/video/526078435_1280x720.jpg',
        description: 'This song is by September 87 and features Dream Fiend.',
        tags: [
            {
                isImportant: true,
                value: 'New Synth Wave',
            },
            {
                isImportant: true,
                value: 'ðŸ”¥Popular',
            },
            {
                value: 'Electronic',
            },
            {
                value: 'Retro',
            },
            {
                value: 'Outrun',
            },
        ],
        comments: [
            {
                user: {
                    name: 'John Wick',
                    avatar:
                        'https://cdnb.artstation.com/p/assets/images/images/013/127/649/large/michael-papatsounis-johnwickfinal.jpg?1538170551',
                },
                text: 'This song is worth killing for',
            },
        ],
    },
    {
        name: 'Running in the night',
        artist: 'FM-84',
        listens: 32000,
        isLiked: false,
        cover: 'https://i.redd.it/xvyfcvt7fvwz.jpg',
        description: 'This song is by September 87 and features Dream Fiend.',
        tags: [
            {
                isImportant: true,
                value: 'New Synth Wave',
            },
            {
                isImportant: true,
                value: 'ðŸ”¥ Popular',
            },
            {
                value: 'Electronic',
            },
            {
                value: 'Retro',
            },
            {
                value: 'Outrun',
            },
        ],
    },
    {
        name: 'Kickstart my heart',
        artist: 'MÃ¶tley CrÃ¼e',
        listens: 1030002,
        isLiked: true,
        cover: 'https://townsquare.media/site/295/files/2019/02/MCscr.jpg?w=980&q=75',
        description: 'This song is by September 87 and features Dream Fiend.',
        tags: [
            {
                value: 'Glam metal',
            },
            {
                isImportant: true,
                value: 'ðŸ”¥ Popular',
            },
            {
                value: 'Heavy metal',
            },
            {
                value: 'Hard Rock',
            },
        ],
    },
    {
        name: 'Dark side of the moon',
        artist: 'Pink Floyd',
        cover: 'https://images-na.ssl-images-amazon.com/images/I/81aTawcGdmL._SL1500_.jpg',
        listens: 8398893,
        isLiked: false,
        description: 'Placeholder',
        tags: [
            {
                value: 'Progressive rock',
                isImportant: true,
            },
            {
                value: 'Rock',
            },
        ],
    },
    {
        name: 'Mind of a Beast',
        artist: 'The Glitch Mob',
        cover: 'https://consequenceofsound.net/wp-content/uploads/2014/01/91vaxjhtjjl-_sl1500_.jpg?quality=80&w=807',
        listens: 490000,
        isLiked: true,
        description: 'Placeholder',
        tags: [
            {
                value: 'Glitch',
                isImportant: true,
            },
            {
                value: 'Electronic rock',
            },
            {
                value: 'Dubstep',
            },
            {
                value: 'EDM',
            },
        ],
    },
];

export class SongsDataSource extends InMemoryDataSource {
    getAll() {
        return songs;
    }
}
