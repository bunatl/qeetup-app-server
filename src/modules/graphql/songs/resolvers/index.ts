import { Resolvers } from '@container/resolvers';
import { queries } from './queries';

export const resolvers: Resolvers = {
    Query: queries,
};
