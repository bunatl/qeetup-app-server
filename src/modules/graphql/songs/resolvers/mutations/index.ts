import { pubsub } from '@container/apollo/pubsub';
import { ResolverObject } from '@container/resolvers';
import { MutationAddCommentArgs, MutationSetLikeArgs } from '@container/schema';
import { Triggers } from '../subscriptions';

export const mutations: ResolverObject = {
    setLike: async (_, { songId, like }: MutationSetLikeArgs, { dataSources: { songs } }) => songs.setLike(songId, like),
    addComment: async (_, { songId, comment }: MutationAddCommentArgs, { dataSources: { songs } }) => {
        await pubsub.publish(Triggers.COMMENT_ADDED, { songId, commentAdded: comment });
        return songs.addComment(songId, comment);
    },
};
