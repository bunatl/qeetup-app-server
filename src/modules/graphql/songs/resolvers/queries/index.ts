import { ResolverObject } from '@container/resolvers';
import { QuerySearchArgs, QuerySongArgs } from '@container/schema';

export const queries: ResolverObject = {
    song: async (_, { id }: QuerySongArgs, { dataSources: { songs } }) => songs.getById(id),
    songs: async (_, __, { dataSources: { songs } }) => songs.getAll(),
    search: async (_, { name }: QuerySearchArgs, { dataSources: { songs } }) => songs.searchByName(name),
};
