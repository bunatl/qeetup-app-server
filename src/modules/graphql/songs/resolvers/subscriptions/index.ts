import { Context } from '@container/apollo';
import { pubsub } from '@container/apollo/pubsub';
import { ResolverObject } from '@container/resolvers';
import { SubscriptionCommentAddedArgs, SubscriptionListensArgs } from '@container/schema';
import { withFilter } from 'apollo-server';

export enum Triggers {
    COMMENT_ADDED = 'COMMENT_ADDED',
    LISTENER_ADDED = 'LISTENER_ADDED',
}

export const subscriptions: ResolverObject = {
    commentAdded: {
        subscribe: withFilter(
            () => pubsub.asyncIterator<Comment>(Triggers.COMMENT_ADDED),
            (payload: { songId: string }, { songId }: SubscriptionCommentAddedArgs) => payload.songId === songId,
        ),
    },
    listens: {
        subscribe: withFilter(
            (_, { songId }: SubscriptionListensArgs, { dataSources: { songs } }: Context) => {
                const iterator = pubsub.asyncIterator<number>(Triggers.LISTENER_ADDED);
                songs.addListener(songId).then((song) => pubsub.publish(Triggers.LISTENER_ADDED, { songId, listens: song.listens }));
                return iterator;
            },
            (payload: { songId: string }, { songId }: SubscriptionListensArgs) => payload.songId === songId,
        ),
    },
};
