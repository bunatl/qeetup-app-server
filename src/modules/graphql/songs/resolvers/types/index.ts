import { ResolverObject } from '@container/resolvers';
import { Song, User } from '@container/schema';

export const types: ResolverObject = {
    Song: {
        comments: (song: Song, _, { dataSources: { songs } }) => songs.getComments(song.id),
    },
    User: {
        __resolveType: (user: User) => (user.isArtist ? 'ArtistUser' : 'RegularUser'),
    },
};
