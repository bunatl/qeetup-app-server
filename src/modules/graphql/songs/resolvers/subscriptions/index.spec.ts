import { pubsub } from '@container/apollo/pubsub';
import { FieldResolver, ResolverObject } from '@container/resolvers';
import { CommentInput, Subscription } from '@container/schema';
import { subscriptions, Triggers } from '.';
import { SongsDataSource } from '../../data-sources';

describe('Songs Subscriptions', () => {
    it('should subscribe to comment adding', async () => {
        const resolver = subscriptions.commentAdded as ResolverObject;
        const subscribe = resolver.subscribe as FieldResolver;

        const songs = new SongsDataSource();
        const allSongs = await songs.getAll();
        const songId = allSongs[0].id;

        // @ts-ignore: Skip info argument.
        const iterator: AsyncIterator<Subscription> = subscribe(null, { songId }, { dataSources: { songs } });

        iterator.next().then(({ value: { commentAdded } }) => expect(commentAdded).toMatchSnapshot());

        const comment: CommentInput = {
            text: 'Test comment',
            user: {
                name: 'test',
                avatar: '...',
            },
        };

        await pubsub.publish(Triggers.COMMENT_ADDED, { songId, commentAdded: comment });
    });

    it('should subscribe to number of listeners', async () => {
        const resolver = subscriptions.listens as ResolverObject;
        const subscribe = resolver.subscribe as FieldResolver;

        const songs = new SongsDataSource();
        const allSongs = await songs.getAll();
        const songId = allSongs[0].id;

        // @ts-ignore: Skip info argument.
        const iterator: AsyncIterator<Subscription> = subscribe(null, { songId }, { dataSources: { songs } });

        iterator.next().then(({ value: { listens } }) => expect(listens).toBe(allSongs[0].listens + 1));
        iterator.next();
    });
});
