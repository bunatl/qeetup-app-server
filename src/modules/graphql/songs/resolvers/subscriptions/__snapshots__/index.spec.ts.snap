// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Songs Subscriptions should subscribe to number of listeners 1`] = `
Object {
  "text": "Test comment",
  "user": Object {
    "avatar": "...",
    "name": "test",
  },
}
`;
